### ISSUE TITLE
Allow SINGLE_LINE_COMMENT to be specified in WhitespaceAfterCheck

### ISSUE THREAD

**Issue by efenderbosch** [checkstyle/checkstyle#6593]
Might just be my personal preference, but I don't like seeing comments like:
```java
//comment goes here
```

I find this:
```java
// easier to read
```

It doesn't even have to be in the default list, just adding it to ``getAcceptableTokens`` would be enough.

````
$ cat TestClass.java
public class TestClass {
//comment goes here
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="WhitespaceAfter">
  <property name="tokens" value="COMMA, SEMI, TYPECAST, LITERAL_IF, LITERAL_ELSE, LITERAL_WHILE, LITERAL_DO, LITERAL_FOR, DO_WHILE" />
</module>
    </module>
</module>

$ java -jar checkstyle-8.18-all.jar -c TestConfig.xml TestClass.java
Starting audit...
Audit done.
````

I would like to see a violation that says a whitespace is needed after `//`.

Code that would produce no violation would be:
````
public class TestClass {
// comment goes here
}
````

Block comments should be supported too, and be configurable separate from single line comments:
````
/*violation here*/
/* no violation here*/
/* no violation here */
````

**Comment by rnveach** [comment]
@romani Please review when you get a chance.

I updated the issue to follow our format.

**Comment by efenderbosch** [comment]
Sorry I didn't follow the format and thanks for the help.

**Comment by romani** [comment]
This Check is not comments aware - https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAfterCheck.java
so simple extension of tokens will not work. To make it "comments aware" is easy update.

Yes , it is very tempting to hack a Check with simple update. But I see that Check already hacked to not a simple space checker after token - https://checkstyle.org/config_whitespace.html#NoWhitespaceAfter , see "allowLineBreaks", so it is focused on java code tokens. Comments tokens were added in some 6.X version of checkstyle.
Base on experience .... I would not recommend to extend java code Check to handle comments. I would rather do new Check for comment tokens and eventually new Check will have comments related properties to adjust it even more. @rnveach , as you saw a lot of time, very universal Checks does not work good, It is better to create new Check and address requirements as separate Check. 

> Might just be my personal preference, but I don't like seeing comments like

Yes, I agree, I do not have strong position on what form to follow and why. But I like comment to be consistent. As I already mentioned comment validation overage is not very strong in Checkstyle, but I would love to extend functionality. BUT I would not recommend to update Java-code Check to handle comments. There are exceptions where Check have to, but your case is not it. Lets keep java-code and comments(javadoc) related Checks separately.
I would not be surprised to get request to enforce "//comment" in method code and  "// comment" between methods. FYI: by comments user can do a lot of code generation, meta content generation, markers, .... and all of them periodically appear with questions no how to make Check to do what they want.

Does it make sense ?

**Comment by rnveach** [comment]
> This Check is not comments aware 

I forgot about this but any check that examines the source file directly is comments aware even if it is specifically enabled. We ran into something like this with indentation and this check is no different and is aware of comments surrounding. See https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAfterCheck.java#L106 and https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAfterCheck.java#L132

I am fine with a new check.

**Comment by romani** [comment]
> I forgot about this but any check that examines the source file directly is comments aware even if it is not specifically enabled.

It is legacy approach, one day we will remove access to FileContent from Checks that already use AST.

---------------

I recommend to create two Checks:
SingleLineCommentFormat
BlockCommentFormat
bother checks will have `regexp` property so user can define what format is expected (with space or without space). Mismatch of regexp will raise violation. Default can be `[ ].*` - start with space.

@efenderbosch  and @rnveach  does it work for you ?
@efenderbosch do you want to contribute such Checks ? we will guide you on how to do this. Such Checks will be easy to do, but whole process might take a bit more time to do all correctly and pass all CIs and review.

**Comment by efenderbosch** [comment]
Yeah, I think that works. I can fork and contribute back w/ some guidance. Sorry for the delay, was on vacation.

**Comment by rnveach** [comment]
> BlockCommentFormat
> regexp 

My only concern is how will the regexp will handle new lines in the block. If you write a regexp that has 2 words next to each other what happens if the block actually has a new line on those 2 words. Will the user have to account for new lines in every place, will we automatically put in a space, or ...?

Lets start with the single line for now.

**Comment by romani** [comment]
yes, single line is more than enough.

@efenderbosch , can you start with opening issue for each Checks separately to discuss details ?
this issue should be closed.

**Comment by ThexXTURBOXx** [comment]
Is there any progress on this? A professor at my university wants us to do exactly this in our code and I always forget about it, which affects my grades.
This feature would help me a lot :)
Is there anything, that I could do to help developing this feature?

**Comment by rnveach** [comment]
Unfortunately nothing has been agreed on for which to even start developing for. See previous comments before yours. To even get something started, we would have to discuss it first and I don't any check being finished in the next month or so.

**Comment by ThexXTURBOXx** [comment]
Oh okay, thank you anyway then :)

**Comment by koppor** [comment]
I think, the follow up is https://github.com/checkstyle/checkstyle/issues/7388.

**Comment by romani** [comment]
@ThexXTURBOXx , you can always help to implement new Check for this - it is not that hard. #7388 already has more ideas on this, there is even workaround (not ideal but it will work to some extend).

I am closing this issue in favor of #7388 , ideas in that ticket are mature.
